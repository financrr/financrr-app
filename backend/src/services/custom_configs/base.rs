use crate::services::custom_configs::bank_data_linking::BankDataLinkingConfig;
use crate::services::custom_configs::opensearch::OpensearchConfig;
use crate::services::Service;
use loco_rs::app::AppContext;
use loco_rs::environment::Environment;
use loco_rs::Error;
use serde::Deserialize;
use std::env;
use std::fs::read_to_string;
use std::sync::{Arc, OnceLock};
use tera::{Context, Tera};
use tracing::{debug, error};

pub const CONFIG_FOLDER_ENV: &str = "LOCO_CONFIG_FOLDER";
pub const DEFAULT_CONFIG_FOLDER: &str = "config";

pub type CustomConfig = Arc<CustomConfigInner>;

#[derive(Debug, Clone, Deserialize)]
pub struct CustomConfigInner {
    pub bank_data_linking: Option<BankDataLinkingConfig>,
    pub opensearch: OpensearchConfig,
}

impl Service for CustomConfigInner {
    async fn new(ctx: &AppContext) -> loco_rs::Result<Self> {
        Self::load_from_env(&ctx.environment)
    }

    fn get_static_once() -> &'static OnceLock<Arc<Self>> {
        static INSTANCE: OnceLock<CustomConfig> = OnceLock::new();

        &INSTANCE
    }
}

impl CustomConfigInner {
    pub fn get_config_folder() -> String {
        env::var(CONFIG_FOLDER_ENV).unwrap_or(DEFAULT_CONFIG_FOLDER.to_string())
    }

    fn load_from_env(env: &Environment) -> loco_rs::Result<Self> {
        let file_name = env.to_string();
        let path = {
            let mut path = Self::get_config_folder();
            if !path.ends_with("/") {
                path.push('/');
            }
            path.push_str(file_name.as_str());
            path.push_str(".yaml");

            path
        };

        debug!("Loading extended config from path: {}", path);

        Self::load_from_path(path)
    }

    fn load_from_path(path: String) -> loco_rs::Result<Self> {
        Self::load_from_string(read_to_string(path.as_str())?, path)
    }

    fn load_from_string(yaml: String, path: String) -> loco_rs::Result<Self> {
        let rendered = Tera::one_off(yaml.as_str(), &Context::new(), false)?;

        serde_yml::from_str(&rendered).map_err(|err| {
            error!("Yaml Error: {} | File: {}", err.to_string(), path);

            Error::Any(err.into())
        })
    }

    pub fn is_bank_data_linking_configured(&self) -> bool {
        if let Some(conf) = &self.bank_data_linking {
            return conf.enabled;
        }

        false
    }
}
