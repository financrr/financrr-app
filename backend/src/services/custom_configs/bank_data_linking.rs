use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct BankDataLinkingConfig {
    #[serde(default = "default_enabled")]
    pub enabled: bool,
    pub implementations: Vec<LinkingImplementation>,
}

#[derive(Debug, Deserialize)]
pub enum LinkingImplementation {
    GoCardless(GoCardlessConfig),
}

impl LinkingImplementation {
    pub fn is_enabled(&self) -> bool {
        match self {
            LinkingImplementation::GoCardless(conf) => conf.enabled,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GoCardlessConfig {
    pub enabled: bool,
    pub secret_id: String,
    pub secret_key: String,
}

fn default_enabled() -> bool {
    true
}
