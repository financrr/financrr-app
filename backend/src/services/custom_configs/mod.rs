pub mod bank_data_linking;
pub mod base;
