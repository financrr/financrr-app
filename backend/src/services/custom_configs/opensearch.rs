use serde::Deserialize;

#[derive(Debug, Clone, Deserialize)]
pub struct OpensearchConfig {
    pub is_https: bool,
    pub host: String,
    pub port: u16,
    pub username: String,
    pub password: String,
    #[serde(default = "default_disable_proxy")]
    pub disable_proxy: bool,
    #[serde(default = "disable_cert_validation")]
    pub disable_cert_validation: bool,
}

impl OpensearchConfig {
    pub fn get_url(&self) -> String {
        let protocol = if self.is_https { "https" } else { "http" };

        format!("{}://{}:{}", protocol, self.host, self.port)
    }
}

pub fn default_disable_proxy() -> bool {
    true
}

pub fn disable_cert_validation() -> bool {
    false
}
