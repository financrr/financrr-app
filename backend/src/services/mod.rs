use crate::services::snowflake_generator::SnowflakeGeneratorInner;
use crate::services::user_verification::UserVerificationServiceInner;
use axum::{Extension, Router as AxumRouter};
use loco_rs::app::AppContext;
use loco_rs::prelude::Result;
use std::future::Future;
use std::sync::{Arc, OnceLock};

pub mod snowflake_generator;
pub mod user_verification;

pub async fn configure_services(router: AxumRouter, ctx: &AppContext) -> Result<AxumRouter> {
    Ok(router
        .layer(UserVerificationServiceInner::get_extension(ctx).await?)
        .layer(SnowflakeGeneratorInner::get_extension(ctx).await?))
}

pub trait Service
where
    Self: Sized + 'static,
{
    fn new(ctx: &AppContext) -> impl Future<Output = Result<Self>>;

    fn get_static_once() -> &'static OnceLock<Arc<Self>>;

    fn new_arc(ctx: &AppContext) -> impl Future<Output = Result<Arc<Self>>> {
        async { Ok(Arc::new(Self::new(ctx).await?)) }
    }

    fn get_arc(ctx: &AppContext) -> impl Future<Output = Result<Arc<Self>>> {
        async {
            let once_lock = Self::get_static_once();

            match once_lock.get() {
                None => {
                    let instance = Self::new_arc(ctx).await?;

                    Ok(once_lock.get_or_init(|| instance).clone())
                }
                Some(instance) => Ok(instance.clone()),
            }
        }
    }

    fn get_extension(ctx: &AppContext) -> impl Future<Output = Result<Extension<Arc<Self>>>> {
        async { Ok(Extension(Self::get_arc(ctx).await?)) }
    }
}
