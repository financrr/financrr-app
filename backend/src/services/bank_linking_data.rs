use crate::bank_account_linking::client::GoCardlessClient;
use crate::services::custom_configs::base::CustomConfigInner;
use crate::services::Service;
use loco_rs::app::AppContext;
use std::sync::{Arc, OnceLock};

pub type BankLinkingData = Arc<BankLinkingDataInner>;

#[derive(Debug, Default)]
pub struct BankLinkingDataInner {
    go_cardless_client: Option<Arc<GoCardlessClient>>,
}

impl Service for BankLinkingDataInner {
    async fn new(ctx: &AppContext) -> loco_rs::Result<Self> {
        let config = CustomConfigInner::get_arc(ctx).await?;
        if !config.is_bank_data_linking_configured() {
            return Ok(BankLinkingDataInner::default());
        }

        if let Some(conf) = config.bank_data_linking.clone() {
            let client = GoCardlessClient::init(conf).await?;

            return Ok(Self {
                go_cardless_client: Some(Arc::new(client)),
            });
        }

        Ok(BankLinkingDataInner::default())
    }

    fn get_static_once() -> &'static OnceLock<Arc<Self>> {
        static INSTANCE: OnceLock<BankLinkingData> = OnceLock::new();

        &INSTANCE
    }
}

impl BankLinkingDataInner {
    pub fn get_go_cardless_client(&self) -> Option<Arc<GoCardlessClient>> {
        self.go_cardless_client.clone()
    }

    pub fn is_go_cardless_client_configured(&self) -> bool {
        match self.get_go_cardless_client() {
            None => false,
            Some(client) => client.is_enabled(),
        }
    }
}
