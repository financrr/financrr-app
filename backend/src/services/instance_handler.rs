use crate::models::_entities::instances;
use crate::services::Service;
use crate::services::snowflake_generator::SnowflakeGeneratorInner;
use loco_rs::Error;
use loco_rs::app::AppContext;
use sea_orm::{DatabaseConnection, IntoActiveModel};
use std::process::abort;
use std::sync::{Arc, OnceLock};
use std::time::Duration;
use tokio_cron_scheduler::{Job, JobScheduler, JobSchedulerError};
use tracing::error;

pub const INSTANCE_HEARTBEAT_INTERVAL_SECONDS: u64 = 10;
pub const INSTANCE_HEARTBEAT_TOLERANCE_SECONDS: u64 = INSTANCE_HEARTBEAT_INTERVAL_SECONDS * 3;

pub type InstanceHandler = Arc<InstanceHandlerInner>;

#[derive(Debug)]
pub struct InstanceHandlerInner {
    instance_id: u16,
}

impl Service for InstanceHandlerInner {
    async fn new(ctx: &AppContext) -> loco_rs::Result<Self> {
        let instance = {
            let instance = instances::Model::get_node_id_and_create_new_instance(&ctx.db).await?;

            SnowflakeGeneratorInner::validate_node_id(instance.node_id)?;

            instance
        };

        let handler = Self::with_node_id(instance.node_id);
        handler
            .start_heartbeat(ctx.db.clone())
            .await
            .map_err(|err| Error::Any(err.into()))?;

        Ok(handler)
    }

    fn get_static_once() -> &'static OnceLock<Arc<Self>> {
        static INSTANCE: OnceLock<Arc<InstanceHandlerInner>> = OnceLock::new();

        &INSTANCE
    }
}

impl InstanceHandlerInner {
    pub fn with_node_id(node_id: i16) -> Self {
        Self {
            instance_id: node_id as u16,
        }
    }

    async fn start_heartbeat(&self, db: DatabaseConnection) -> loco_rs::Result<(), JobSchedulerError> {
        let scheduler = JobScheduler::new().await?;

        let node_id = self.instance_id as i16;
        let job = Job::new_repeated_async(
            Duration::from_secs(INSTANCE_HEARTBEAT_INTERVAL_SECONDS),
            move |_uuid, _l| {
                let db = db.clone();
                Box::pin(async move {
                    match instances::Model::find_by_node_id(&db, node_id).await {
                        Ok(instance) => {
                            let active_model = instance.into_active_model();
                            if let Err(e) = active_model.update_heartbeat(&db).await {
                                error!("Failed to update heartbeat: {}", e);
                                abort()
                            }
                        }
                        Err(err) => {
                            error!(
                                "Failed to get instance by current node id: {} | Error: {}",
                                node_id, err
                            );
                            abort()
                        }
                    }
                })
            },
        )?;

        scheduler.add(job).await?;

        scheduler.shutdown_on_ctrl_c();
        scheduler.start().await?;

        Ok(())
    }

    pub fn get_instance_id(&self) -> u16 {
        self.instance_id
    }
}
