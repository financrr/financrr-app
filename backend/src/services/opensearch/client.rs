use crate::error::app_error::AppResult;
use crate::services::custom_configs::base::CustomConfigInner;
use crate::services::custom_configs::opensearch::OpensearchConfig;
use crate::services::Service;
use loco_rs::app::AppContext;
use opensearch::http::transport::{SingleNodeConnectionPool, TransportBuilder};
use opensearch::http::Url;
use opensearch::OpenSearch;
use std::sync::{Arc, OnceLock};

pub type OpensearchClient = Arc<OpensearchClientInner>;

pub struct OpensearchClientInner {
    config: OpensearchConfig,
    opensearch: OpenSearch,
}

impl Service for OpensearchClientInner {
    async fn new(ctx: &AppContext) -> loco_rs::Result<Self> {
        let custom_config = CustomConfigInner::get_arc(ctx).await?;

        let internal_client = Self::get_client(&custom_config.opensearch).await?;

        Ok(Self {
            config: custom_config.opensearch.clone(),
            opensearch: internal_client,
        })
    }

    fn get_static_once() -> &'static OnceLock<Arc<Self>> {
        static INSTANCE: OnceLock<OpensearchClient> = OnceLock::new();

        &INSTANCE
    }
}

impl OpensearchClientInner {
    async fn get_client(config: &OpensearchConfig) -> AppResult<OpenSearch> {
        let url = Url::parse(config.get_url().as_str())?;
        let pool = SingleNodeConnectionPool::new(url);

        let mut transport = TransportBuilder::new(pool);
        if config.disable_proxy {
            transport = transport.disable_proxy();
        }
        let transport = transport.build()?;

        Ok(OpenSearch::new(transport))
    }
}
