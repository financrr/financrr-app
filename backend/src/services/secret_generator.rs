use crate::services::instance_handler::{InstanceHandler, InstanceHandlerInner};
use crate::services::Service;
use base64::prelude::BASE64_URL_SAFE;
use base64::Engine;
use loco_rs::app::AppContext;
use rand::rngs::OsRng;
use rand::RngCore;
use sha2::{Digest, Sha256};
use std::sync::{Arc, OnceLock};

pub const DEFAULT_TOKEN_LENGTH: usize = 64;

pub type SecretGenerator = Arc<SecretGeneratorInner>;

pub struct SecretGeneratorInner {
    instance_handler: InstanceHandler,
}

impl Service for SecretGeneratorInner {
    async fn new(ctx: &AppContext) -> loco_rs::Result<Self> {
        Ok(Self {
            instance_handler: InstanceHandlerInner::get_arc(ctx).await?,
        })
    }

    fn get_static_once() -> &'static OnceLock<Arc<Self>> {
        static INSTANCE: OnceLock<Arc<SecretGeneratorInner>> = OnceLock::new();

        &INSTANCE
    }
}

impl SecretGeneratorInner {
    /// Generates a random token with a default length of 64 characters
    pub fn generate_token(&self) -> String {
        self.generate_token_with_length(DEFAULT_TOKEN_LENGTH)
    }

    pub fn generate_token_with_length(&self, length: usize) -> String {
        let mut random_bytes = [0u8; 32];
        OsRng.fill_bytes(&mut random_bytes);

        let mut hasher = Sha256::new();
        hasher.update(self.instance_handler.get_instance_id().to_string().as_bytes());
        hasher.update(random_bytes);
        let hash = hasher.finalize();

        BASE64_URL_SAFE.encode(hash).chars().take(length).collect()
    }
}
