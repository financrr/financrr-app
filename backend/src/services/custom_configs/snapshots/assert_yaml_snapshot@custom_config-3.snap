---
source: src/services/custom_configs/base.rs
assertion_line: 217
expression: config
---
CustomConfigInner {
    bank_data_linking: Some(
        BankDataLinkingConfig {
            enabled: true,
            secret_id: "test_id",
            secret_key: "test_key",
            api_url: "https://bankaccountdata.gocardless.com/",
        },
    ),
}
